lib/spack/llnl/util/cpu.py

```diff
@@ -0,0 +1,328 @@
+
+        targets[name] = MicroArchitecture(
+            name, parents, vendor, features, compilers, generation
+        )
+
+    this_dir = os.path.dirname(os.path.abspath(__file__))
+    filename = os.path.join(this_dir, 'targets.json')
+    with open(filename, 'r') as f:
+        data = json.load(f)
+
+    targets = OrderedDict()
+    for name in data:
+        if name in targets:
+            # name was already brought in as ancestor to a target
+            continue
+        fill_target_from_dict(name, data, targets)
+
+    # Add the host platform if not present
+    host_platform = platform.machine()
+    targets.setdefault(host_platform, generic_microarchitecture(host_platform))
+
+    return targets
+
+
+#: Dictionary of known micro-architectures
+targets = _load_microarchitectures_from_json()
+
+
+def supported_target_names():
+    return targets.keys()
+
+
+def _create_cpuinfo_dict():
+    """Returns a dictionary with information on the host CPU."""
+    dict_factory = {
+        'Linux': _create_dict_from_proc,
+        'Darwin': _create_dict_from_sysctl
+    }
+    return dict_factory[platform.system()]()
+
+
+def _create_dict_from_proc():
+    # Initialize cpuinfo from file
+    cpuinfo = {}
+    try:
+        with open('/proc/cpuinfo') as file:
+            text = file.readlines()
+            for line in text:
+                if line.strip():
+                    key, _, value = line.partition(':')
+                    cpuinfo[key.strip()] = value.strip()
+    except IOError:
+        return None
+    return cpuinfo
+
+
+def _create_dict_from_sysctl():
+
+    def check_output(args):
+        if sys.version_info >= (3, 0):
+            return subprocess.run(
+                args, check=True, stdout=subprocess.PIPE).stdout  # nopyqver
+        else:
+            return subprocess.check_output(args)  # nopyqver
+
+    cpuinfo = {}
+    try:
+        cpuinfo['vendor_id'] = check_output(
+            ['sysctl', '-n', 'machdep.cpu.vendor']
+        ).strip()
+        cpuinfo['flags'] = check_output(
+            ['sysctl', '-n', 'machdep.cpu.features']
+        ).strip().lower()
+        cpuinfo['flags'] += ' ' + check_output(
+            ['sysctl', '-n', 'machdep.cpu.leaf7_features']
+        ).strip().lower()
+        cpuinfo['model'] = check_output(
+            ['sysctl', '-n', 'machdep.cpu.model']
+        ).strip()
+        cpuinfo['model name'] = check_output(
+            ['sysctl', '-n', 'machdep.cpu.brand_string']
+        ).strip()
+
+        # Super hacky way to deal with slight representation differences
+        # Would be better to somehow consider these "identical"
+        if 'sse4.1' in cpuinfo['flags']:
+            cpuinfo['flags'] += ' sse4_1'
+        if 'sse4.2' in cpuinfo['flags']:
+            cpuinfo['flags'] += ' sse4_2'
+        if 'avx1.0' in cpuinfo['flags']:
+            cpuinfo['flags'] += ' avx'
+    except Exception:
+        pass
+    return cpuinfo
+
+
+def detect_host():
```

Comment:
I think `detect_host()` should return *one* value, and there should be another `compatible_hosts()` function.

Also they should both probably be called `detect_march()` and `compatible_marches()` or something like that.  Hmm... `marches` doesn't read very easily does it?

Reply:
At the moment I didn't add a `compatible_hosts` function, but in case we need it it's simply:
```python
def compatible_hosts():
    host = detect_host()
    return [host] + host.ancestors
```